flashinfer/cute_dsl/add_rmsnorm_fp4quant.py

```diff
@@ -0,0 +1,2278 @@
+"""
+Copyright (c) 2025 by FlashInfer team.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Fused Add + RMSNorm + FP4 Quantization using CuTe-DSL
+======================================================
+
+High-performance fused kernel for element-wise addition followed by RMS normalization
+and FP4 quantization. Supports both NVFP4 (block_size=16, E4M3 scales) and MXFP4
+(block_size=32, UE8M0 scales) formats.
+
+"""
+
+import functools
+import math
+import operator
+from typing import Callable, Tuple
+
+import cuda.bindings.driver as cuda
+import cutlass
+import cutlass.cute as cute
+import cutlass.torch as cutlass_torch
+import torch
+from cutlass import Float32, Int32, Int64, Uint32, Uint64, Uint8
+from cutlass.cutlass_dsl import T, dsl_user_op
+from cutlass._mlir.dialects import llvm
+
+from .utils import make_ptr
+from ..api_logging import flashinfer_api
+
+# =============================================================================
+# Constants
+# =============================================================================
+
+FLOAT4_E2M1_MAX = 6.0
+FLOAT8_E4M3_MAX = 448.0
+SF_VEC_SIZE = 16
+COPY_BITS = 128
+
+
+# =============================================================================
+# Architecture Detection
+# =============================================================================
+
+
+@functools.cache
+def get_sm_version() -> int:
```

Comment:
How about passing a torch.device to this function? (In case user is operating GPU1 when current device is GPU 0)